File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int sum=Integer.MIN_VALUE;
    //int maxsum=0;
    public int maxPathSum(TreeNode root) {
        height(root);
        return sum;
    }
    private int height(TreeNode root){
        if(root==null){
            return 0;
        }
        int leftsum=Math.max(height(root.left),0);
        int rightsum=Math.max(height(root.right),0);

         sum=Math.max(sum,leftsum+rightsum+root.val);

        return Math.max(leftsum,rightsum)+root.val;
    }
}

File: 124-binary-tree-maximum-path-sum/binary-tree-maximum-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        
        res=[root.val]
        def pathsum(root):
            if root is None:
                return 0
            
            leftMax= max(0, pathsum(root.left))
            rightMax=max(0, pathsum(root.right))

            res[0]= max(leftMax+rightMax+root.val, res[0])

            return root.val+ max(leftMax,rightMax )
        pathsum(root)
        return res[0]